File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        //All Ascii chars
        int HashLen = 256;
        //hashtable to store last occurence of each char
        int[] hash = new int[HashLen];
        //Initialize the table with -1
        Arrays.fill(hash,-1);
        int l=0, r=0, maxLen=0;
        while(r<n){
            //If current character s.charAt(r) is already in the substring
            if(hash[s.charAt(r)] >= l){
                //Move left pointer to the right of the last occurence of s.charAt(r)
            l= Math.max(hash[s.charAt(r)]+1, l);
            }
            int len = r-l+1;
            maxLen= Math.max(len, maxLen);
            //store the index of current chatacter in the hash table
            hash[s.charAt(r)] =r;
            r++;
        }
        return maxLen;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        n= len(s)
        HashLen = 256
        """ Hash table to store last
            occurrence of each character """
        hash = [-1] * HashLen
        for i in range(HashLen):
            hash[i] = -1
        l,r, maxLen = 0,0,0
        while r< n:
            """ If current character s[r] 
                is already in the substring    """
            if hash[ord(s[r])] !=-1:
                l = max(hash[ord(s[r])] +1, l)
            current_len = r-l+1
            maxLen= max(current_len, maxLen)
            #Store the index of the currentcharacter in the hash table
            hash[ord(s[r])] = r   
            # Move right pointer to next position
            r += 1
        
        # Return the maximum length found
        return maxLen       